include/deal.II/integrators/advection.h

```diff
@@ -311,51 +516,198 @@ namespace LocalIntegrators
-                    }
-                  else
-                    {
-                      M22(i,j) -= dx_nbeta
-                                  * fe2.shape_value(j,k)
-                                  * fetest2.shape_value(i,k);
-                      M12(i,j) += dx_nbeta
-                                  * fe2.shape_value(j,k)
-                                  * fetest1.shape_value(i,k);
-                    }
-                }
-              else
-                {
-                  for (unsigned int d=0; d<fe1.get_fe().n_components(); ++d)
+              {
+                if (fe1.get_fe().is_primitive())
+                  {
+                    const double v1 = fetest1.shape_value(i,k);
+                    const double v2 = fetest2.shape_value(i,k);
+                    const double u1 = fe1.shape_value(j,k);
+                    const double u2 = fe2.shape_value(j,k);
                     if (nbeta > 0)
                       {
-                        M11(i,j) += dx_nbeta
-                                    * fe1.shape_value_component(j,k,d)
-                                    * fetest1.shape_value_component(i,k,d);
-                        M21(i,j) -= dx_nbeta
-                                    * fe1.shape_value_component(j,k,d)
-                                    * fetest2.shape_value_component(i,k,d);
+                        M11(i,j) += dx_nbeta*u1*v1;
+                        M21(i,j) -= dx_nbeta*u1*v2;
                       }
                     else
                       {
-                        M22(i,j) -= dx_nbeta
-                                    * fe2.shape_value_component(j,k,d)
-                                    * fetest2.shape_value_component(i,k,d);
-                        M12(i,j) += dx_nbeta
-                                    * fe2.shape_value_component(j,k,d)
-                                    * fetest1.shape_value_component(i,k,d);
+                        M12(i,j) += dx_nbeta*u2*v1;
+                        M22(i,j) -= dx_nbeta*u2*v2;
+                      }
+                  }
+                else
+                  {
+                    for (unsigned int d=0; d<fe1.get_fe().n_components(); ++d)
+                      {
+                        const double v1 = fetest1.shape_value_component(i,k,d);
+                        const double v2 = fetest2.shape_value_component(i,k,d);
+                        const double u1 = fe1.shape_value_component(j,k,d);
+                        const double u2 = fe2.shape_value_component(j,k,d);
+                        if (nbeta > 0)
+                          {
+                            M11(i,j) += dx_nbeta*u1*v1;
+                            M21(i,j) -= dx_nbeta*u1*v2;
+                          }
+                        else
+                          {
+                            M12(i,j) += dx_nbeta*u2*v1;
+                            M22(i,j) -= dx_nbeta*u2*v2;
+                          }
```

Comment:
If we care about efficiency here, I'd like to get rid of `nbeta > 0` here and get it out of the loop over i and j. You would then create two references to matrices
```
FullMatrix<double> &M1 = nbeta > 0 ? M11 : M12;
FullMatrix<double> &M2 = nbeta > 0 ? M21 : M22;
const FEValuesBase<dim,spacedim> &fe = neta >  ? fe1 : fe2;
```
and then run the loop without the if (conditional branch) inside. You might argue that a compiler or branch predictor will take care of it, but my experience is that a compiler cannot do it because there is not enough aliasing information; consequently, reducing the optimization possibilities over several matrix items. I would expect gains somewhere between 5 and 20 percent, depending on the compiler and CPU.